File: acme-admin/src/main/java/com/acme/admin/controller/ProductOrServiceController.java
package com.acme.admin.controller;

import com.acme.admin.dto.ProductOrServiceCreateRequest;
import com.acme.admin.dto.ProductOrServiceResponse;
import com.acme.admin.service.ProductOrServiceService;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/catalog/items")
public class ProductOrServiceController {
  private final ProductOrServiceService service;

  public ProductOrServiceController(ProductOrServiceService service) {
    this.service = service;
  }

  @GetMapping
  @PreAuthorize("hasAuthority('CATALOG_READ')")
  public List<ProductOrServiceResponse> list() {
    return service.list();
  }

  @PostMapping
  @PreAuthorize("hasAuthority('CATALOG_WRITE')")
  public ProductOrServiceResponse create(@Valid @RequestBody ProductOrServiceCreateRequest req) {
    return service.create(req);
  }
}

File: acme-admin-dotnet/Controllers/ProductOrServiceController.cs
using Acme.Admin.Api.DTO;
using Acme.Admin.Api.Security;
using Acme.Admin.Api.Services;
using Acme.Admin.Api.Validation;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Acme.Admin.Api.Controllers;

[ApiController]
[Route("api/catalog/items")]
public sealed class ProductOrServiceController(ProductOrServiceService service) : ControllerBase
{
    [HttpGet]
    [Authorize(Policy = PermissionPolicies.CatalogRead)]
    public async Task<ActionResult<PagedResult<ProductOrServiceResponse>>> List(
        [FromQuery] int page = Paging.DefaultPage,
        [FromQuery] int size = Paging.DefaultSize,
        CancellationToken ct = default)
    {
        var (p, s) = Paging.Normalize(page, size);
        return Ok(await service.ListItemsAsync(p, s, ct));
    }

    [HttpPost]
    [Authorize(Policy = PermissionPolicies.CatalogWrite)]
    public async Task<ActionResult<ProductOrServiceResponse>> Create([FromBody] ProductOrServiceCreateRequest req, CancellationToken ct)
    {
        return Ok(await service.CreateItemAsync(req, ct));
    }
}
